### PR TITLE
ENH: optimize.elementwise: vectorized scalar optimization and root finding tools

### LINKED ISSUE #7242: ENH: implement at least one good robust scalar root-finding algorithm (Brent's? Chandrupatla's?) in a vectorized approach
I'm deep in the thick of some Monte Carlo analysis using a root-finding algorithm for each sample. For performance reasons I would like to use batch-processing techniques, but [`scipy.optimize.brentq`](https://docs.scipy.org/doc/scipy-0.13.0/reference/generated/scipy.optimize.brentq.html#scipy.optimize.brentq) operates only on scalars, not vectors.

Problem statement:
- Given some function `f` of at least one argument, which can operate on vectors for at least its first argument, in which case it returns a vector of the same size as the first argument
- Create a function `rootsolver` (or other appropriate name) which can be called:
    
      rootsolver(f,a,b,xtol,rtol,maxiter,args,...)

where a,b,xtol,rtol,maxiter,args have the same meaning as in `scipy.optimize.brentq` but all of them except maxiter and args can be *vectors* (1-D arrays) as long as they are of identical length, and in this case `rootsolver` proceeds in a vectorized manner:

  - choosing initial vector guesses for `x`
  - calling `y=f((x,)+args)` repeatedly
  - choosing updated vector guesses for `x` on each iteration, based on current and recent values `y`
  - terminating when *all* of the values `x` are within sufficient tolerance, or when maxiter is reached
  - if a count of iterations is returned (e.g. as in `RootResults`) it should be a vector of such counts

This requires if-statements to be turned into vectorized equivalents (e.g. conditional assignment based on boolean arrays)

Simple test case:

```
import numpy as np
from scipy.optimize import brentq
def f(x,a):
    return a-x*x

brentq(f,0,3,args=(2,))   # returns 1.4142135623731364
k=np.arange(1,8)
brentq(f,0,3,args=(k,))   # should return an array which is approximately sqrt(k)
```

Ideally this would "play well with numba" but I see that the `scipy.optimize` algorithms are already implemented in C so maybe that isn't a reasonable request.

### LINKED ISSUE #18445: ENH: add a callback feature to minimize_scalar and root_scalar
In scipy.optimize.minimize_scalar(), there is an option to specify `xtol` (or `xatol` in the `bounded` method) in the `options` dictionary.   Would be nice to have a input that specifies a tolerance on the output such that when the optimizer approaches close to the result it concludes the optimization. 

ex. say you want to find the minimum of `f(x)`. The minimum is 0. Maybe I don't need the optimizer to drill down `f(x)` all the way down to say 1e-3, I'm satisfied with getting only as close as 1e-1 for my use because `f(x)` is a computationally expensive function and maybe I don't need that much accuracy

example code below, if my explanation is not clear:
```
def fun(x0):
    return x**2 -1

res = scipy.optimize.minimize_scalar(fun,options={"output_tol:1e-1"}
```
In this case `output_tol` would stop the optimizer when `res.fun` is within (-0.1,0.1). 

Edit: fixed errors pointed out by tupui

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-7242
Closes gh-18445
Toward gh-6379
Toward gh-18867

#### What does this implement/fix?
There have been requests for scalar rootfinders and minimizers to have a callback interface and be vectorized like `newton`. It would also be nice for such rootfinders and minimizers to have a consistent interface, accept both absolute and relative tolerances on the abscissa and function value, provide detailed output information, and be array-API compatible. This PR attempts to address these needs by providing a public interface to the following private functions:

- [`scipy.optimize._chandrupatla._chandrupatla`](https://github.com/scipy/scipy/blob/00e8c70ae21013c7822a7b5b1fd7b74ff260373b/scipy/optimize/_chandrupatla.py#L12) (rootfinding) → `scipy.optimize.elementwise.rootfind`
- [`scipy.optimize._chandrupatla._chandrupatla_minimize`](https://github.com/scipy/scipy/blob/00e8c70ae21013c7822a7b5b1fd7b74ff260373b/scipy/optimize/_chandrupatla.py#L274) (minimization) → `scipy.optimize.elementwise.minimize`
- [`scipy.optimize._bracket._bracket_root`](https://github.com/scipy/scipy/blob/00e8c70ae21013c7822a7b5b1fd7b74ff260373b/scipy/optimize/_bracket.py#L52) (root bracketing) → `scipy.optimize.elementwise.bracket_root`
- [`scipy.optimize._bracket._bracket_minimum`](https://github.com/scipy/scipy/blob/00e8c70ae21013c7822a7b5b1fd7b74ff260373b/scipy/optimize/_bracket.py#L458) (minimum bracketing) → `scipy.optimize.elementwise.bracket_minimum`

Rationale for adding new functions rather than extending `root_scalar` and `root_minimize` can be found in https://github.com/scipy/scipy/issues/7242#issuecomment-2060485869 and https://github.com/scipy/scipy/pull/20624#issue-2274635560, respectively. These functions are put in a new `elementwise` namespace rather than the base `optimize` namespace to keep the function names manageable without conflicting with existing function names. The APIs for `rootfind` and `minimize` have been designed with the option to add a `method` argument in mind; they will be equally appropiate for other common bracketing and derivative-based rootfinding and minimization methods.

#### Additional information
Currently, only `elementwise.rootfind` has array API support (see gh-20689); the rest will be converted over the summer. 

Questions:
1. The private functions are extensively tested in [`test_chandrupatla.py`](https://github.com/scipy/scipy/blob/main/scipy/optimize/tests/test_chandrupatla.py) and [`test_bracket.py`](https://github.com/scipy/scipy/blob/main/scipy/optimize/tests/test_bracket.py). In this PR, I don't want to convert the tests completely to use the new interfaces; this can be done when new methods are added and the tests are parameterized over all methods. For this PR, how about we leave the tests untouched but a) rather than importing the private functions in the test file, we import only the public functions and b) include in the test file a wrapper that exposes the public function with the interface of the private function. (Yes, effectively, the arguments and returns would get converted back and forth.) This would keep the diff tiny so that this PR can focus on the API rather than the nitty gritty. *Update: Done.*
2. I've been careful to use "root" to refer to the abscissa at which the function value is "zero", and sometimes I am careful to make a similar distinction between "minimizer" and "minimum". From this perspective, the function name `bracket_minimum` might not be consistent with `bracket_root`; perhaps `bracket_minimum` should be `bracket_minimizer` or `bracket_root` should be `bracket_zero`. This is worth discussing. *Update: "minimum" is acceptable. "minimizer" could be confused with an algorithm or function for performing minimization.*
3. Should we mark arrays as read-only before passing them into user callables? I'd been thinking about it already, and it came up recently in gh-20803. They might already be copies, in which case it wouldn't matter, but it might be good anyway. *Update: I don't think there is an array API standard way to do this, so I've documented that callables must not mutate the arrays.*

To do:
- [x] More extensive examples
- [x] Note conditions under which functions will converge successfully
- [x] Add tests according to the resolution of question 1

### CODE REVIEW

**Comment by tupui** [comment]
For the naming I think I would do
```
elementwise.find_root
elementwise.bracket_root
elementwise.bracket_minimum
elementwise.minimum
```

Or use verbs for the above to have `elementwise.minimize`, `elementwise.bracket_minimumizer`

**Comment by mdhaber** [comment]
It sounds like you are suggesting to change from verbs to nouns, but then at the end you write that verbs are OK? They are all intended to be verbs right now. The one that is questionable is "rootfind". "root-finding" is a present participle verb that is often used as an adjective, suggesting that "root-find" could be used as a verb. It's not common to do so in English, but there are a fair number of functions named `rootfind` that have a similar purpose. To better address the comment, what is the intent of the suggestions?

**Comment by tupui** [comment]
I meant that I would either use nouns or verbs. Not mix.

Also this could work
```
elementwise.root_finder
elementwise.bracket_root_finder
elementwise.minimizer
elementwise.bracket_minimizer
```

**Comment by mdhaber** [comment]
> I meant that I would either use nouns or verbs. Not mix.

Right. I interpret the existing names as verbs. Verb or noun is an age-old debate, and we are inconsistent in `optimize` (verb); e.g., we have `minimize` and `bracket` (verbs), but also `root` and `linprog` (noun/nounish).

"minimize" is a verb. "bracket" is a verb or a noun, but it seems you can interpret it as a verb, too, given the suggestion of `bracket_minimumizer` above. So `rootfind` seems to be the one in question. It was intended to be a verb in the sense that "root" is just a noun that modifies the verb "find", but if group prefers for the name to start with `find`, OK.

**Comment by tupui** [comment]
`bracket_minimizer` yeah this one I just confused myself and did not think about bracket as a verb even. 

For root and find, I think that it would be better to harmonise it with minimize as I proposed above (modulo the grammar). So if we have a `bracket_minimum` and `minimum` then seems normal to me to have `bracket_root` with `root`

**Comment by mdhaber** [comment]
I interpret that as mixing verbs and nouns, though. There, your `minimum` and `root` are nouns, but `bracket_root` and `bracket_minimum` seem to be verbs because of the word order, e.g. "to bracket a root". `minimum_bracket` and `root_bracket` would be much easier to interpret as nouns.

**Comment by tupui** [comment]
Yeah I mean, `braketer_root` maybe

**Comment by mdhaber** [comment]
This is getting quite long, so this will be my last post about the names until we hear from others.

Assuming by `braketer_root` you mean `bracketer_root`, this would be a noun. But I think it goes against the principles of why a function might be named with a noun instead of a verb - to identify what the function returns. The function itself could be said to be a "root bracketer", but it _returns_ a "bracket", not a "bracketer".

**Comment by steppi** [comment]
> As a compromise, I can live with `find` as the verb for both `root` and `minimum`/`minimizer`.

I prefer `minimize`, my preferred order is

$$\mathrm{minimize} > \mathrm{find\\_minimum} > \mathrm{find\\_minimizer}$$

It just so happens that there isn't a good verb to describe finding roots of a functions, but I don't think that means we should abandon a good verb for finding the minimum/minimizer. I like using `minimum` over `minimizer` due to an ambiguity between `minimizer` as the `x` value where a `minimum` is attained, and an algorithm which finds this `x` value. Similarly, for the related bracketing function, I like `bracket_minimum`. I think `bracket_minimizer` might even be interpreted as a bracketing minimization algorithm by some people.

I like `bracket_root` for the function which finds an initial bracket for the root. I think I prefer `find_root` over `rootfind`, but don't have a strong opinion.

**Comment by mdhaber** [comment]
OK, so with `minimizer` out of the way, we're looking at:
```python3
- `find_root`
- `find_minimum` vs `minimize`
- `bracket_root`
- `bracket_minimum`
```

**Comment by tupui** [comment]
My vote is on
```
find_root
find_minimum
bracket_root
bracket_minimum
```
Symmetric for users

**Comment by steppi** [comment]
I can accept `find_minimum`.

**Comment by mdhaber** [comment]
Great. How about questions 1 and 3?

**Comment by tupui** [comment]
Yes to both for me.

I prefer to keep this focused and for the arrays I personally preferred when my data is not modified by the function using it.

**Comment by steppi** [comment]
> Great. How about questions 1 and 3?

For question 1) +1 for keeping the diff small by writing a wrapper like you described.
For question 2) I think marking arrays read-only is the principled way to do it, so I'm +1 for that too. I assume requiring user supplied functions written in Cython to have arguments marked `const` shouldn't be an issue here right, since there's no backwards compatibility to worry about?

**Comment by mdhaber** [comment]
> I assume requiring user supplied functions written in Cython to have arguments marked `const` shouldn't be an issue here right, since there's no backwards compatibility to worry about?

Right.

> I prefer to keep this focused

I assume you mean that this should be a separate PR.  Probably - I think it can/should be changed in the EIM infrastructure anyway, not this code.

I'll add the tests.

**Comment by mdhaber** [comment]
I'm not sure why the CircleCI doc build job is not showing up above, but the link is [here](https://app.circleci.com/pipelines/github/mdhaber/scipy/506/workflows/7f4b94b6-1568-4503-b695-d04d9d3e2c81/jobs/1825). As you can see in the [rendered docs](https://output.circle-artifacts.com/output/job/9f4fd92d-af92-4f87-8057-090a14fde8b4/artifacts/0/html/reference/optimize.html#elementwise-minimization-and-rootfinding), I'm not getting the expected TOC from `elementwise.py` and Sphinx is reporting "WARNING: toctree contains reference to nonexisting document 'reference/optimize.elementwise'". I'm trying to find the examples of `scipy.stats.sampling`, `qmc`, `contingency`, and I think what I have is analogous - do you know how to fix this?

**Comment by mdhaber** [comment]
6565ce27a24b2d4fbc67f1e9e9fcedb9bab15775 demonstrates what I mean for the tests, item 1. LMK if that looks good and I'll repeat that for the other functions.

**Comment by mdhaber** [comment]
I'll try to fix the failure after the doc build finishes.
```
________________________ test_all_modules_are_expected _________________________
scipy/_lib/tests/test_public_api.py:255: in test_all_modules_are_expected
    raise AssertionError(f'Found unexpected modules: {modnames}')
E   AssertionError: Found unexpected modules: ['scipy.optimize.elementwise']
        _          = False
        ignore_errors = <function test_all_modules_are_expected.<locals>.ignore_errors at 0x144ca0b80>
        modname    = 'scipy.version'
        modnames   = ['scipy.optimize.elementwise']
```

File: scipy/optimize/__init__.py [hunk 1]
```diff
@@ -240,6 +240,14 @@
    optimize.root-excitingmixing
    optimize.root-krylov
    optimize.root-dfsane
+   
+Elementwise Minimization and Rootfinding
+========================================
```
Comment (mdhaber) [inline scipy/optimize/__init__.py:245]:
```suggestion
Elementwise Minimization and Root Finding
=========================================
```

File: scipy/optimize/elementwise.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+"""
+===================================================================
+Elementwise Scalar Optimization (:mod:`scipy.optimize.elementwise`)
+===================================================================
+
+.. currentmodule:: scipy.optimize.elementwise
+
+This module provides a collection of functions for rootfinding and
```
Comment (mdhaber) [inline scipy/optimize/elementwise.py:8]:
```suggestion
This module provides a collection of functions for root finding and
```

File: scipy/optimize/elementwise.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+minimization of scalar, real-valued functions of one variable. Unlike their
+counterparts in the base :mod:`scipy.optimize` namespace, these functions work
+elementwise, enabling the solution of many related problems in an efficient,
+vectorized call. Furthermore, when environment variable ``SCIPY_ARRAY_API=1``,
+these functions can accept non-NumPy, array API standard compatible arrays and
+perform all calculations using the corresponding array library (e.g. PyTorch,
+JAX, CuPy).
+
+Rootfinding
+===========
```
Comment (mdhaber) [inline scipy/optimize/elementwise.py:18]:
```suggestion
Root finding
============
```

File: scipy/optimize/elementwise.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+.. autosummary::
+   :toctree: generated/
+
+   find_root
+   bracket_root
+
+Rootfinding
+===========
```
Comment (mdhaber) [inline scipy/optimize/elementwise.py:27]:
```suggestion
Minimization
============
```

File: scipy/optimize/_elementwise.py [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+    else:
+        _callback = callback
+
+    res = _chandrupatla(f, xl, xr, args=args, **tolerances,
+                        maxiter=maxiter, callback=_callback)
+    return reformat_result(res)
+
+
+def find_minimum(f, init, /, *, args=(), tolerances=None, maxiter=100, callback=None):
+    """Find the minimizer of an unimodal, real-valued function of a real variable.
```
Comment (mdhaber) [inline scipy/optimize/_elementwise.py:235]:
```suggestion
    """Find the minimum of an unimodal, real-valued function of a real variable.
```

**Review by tupui — APPROVED** [review]
Thanks Matt. This has gone through the forum and LGTM 🚀

**Pull request merged**